File: src/ASTComponents/ExpressionComponents/operator_expression.rb

class OperatorExpresison
    def initialize(token, lhs_exp, rhs_exp)
        @lhs = lhs_exp
        @rhs = rhs_exp
        @type = token.getType()
        @token = token
    end

    def visit(semantic_analyzer)
        semantic_analyzer.analyze_node_locally(self)
    end

    def getRhsExp()
        return @rhs
    end

    def getLhsExp()
        return @lhs
    end

    def getName()
        return getOperator()
    end

    def getOperator()
        return @token
    end

    def _printLiteral(repr_list)
        repr_list.append('(')
        @lhs._printLiteral(repr_list)
        repr_list.append(' ' + @token.getText() + ' ')
        @rhs._printLiteral(repr_list)
        repr_list.append(')')
    end

    def _printTokType(type_list)
        type_list.append('(')
        @lhs._printTokType(type_list)
        type_list.append(' ' + @type.to_s + ' ')
        @rhs._printTokType(type_list)
        type_list.append(')')
    end

    def toJSON()
        return {
            "type" => "binary",
            "token" => {
                "literal" => @token.getText(),
                "type" => @token.getType(),
                "line_number" => @token.getLine()
            },
            "lhs_exp" => @lhs.toJSON(),
            "rhs_exp" => @rhs.toJSON()
        }
    end
end

File: src/ASTComponents/ExpressionComponents/operator_expression.py


class OperatorExpression:
    def __init__(self):
        self.operator_token = None
        self.left_expression = None
        self.right_expression = None

    def add_name(self, operator_token):
        self.operator_token = operator_token

    def get_name(self):
        return self.operator_token
    
    def add_lhs_exp(self, lhs_exp):
        self.left_expression = lhs_exp

    def get_lhs_exp(self):
        return self.left_expression
    
    def add_rhs_exp(self, rhs_exp):
        self.right_expression = rhs_exp

    def get_rhs_exp(self):
        return self.right_expression

    def accept(self, expression_analyzer):
        expression_analyzer.visit_binary_operator_expression(self)
